node-maintainer/src/main/java/com/yahoo/vespa/hosted/node/certificate/CertificateAuthorityClient.java

```diff
@@ -0,0 +1,92 @@
+package com.yahoo.vespa.hosted.node.certificate;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.yahoo.io.IOUtils;
+import com.yahoo.vespa.hosted.athenz.instanceproviderservice.ca.model.CertificateSerializedPayload;
+import com.yahoo.vespa.hosted.athenz.instanceproviderservice.ca.model.CsrSerializedPayload;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.conn.ssl.SSLConnectionSocketFactory;
+import org.apache.http.conn.ssl.SSLContextBuilder;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.bouncycastle.pkcs.PKCS10CertificationRequest;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.security.cert.X509Certificate;
+
+/**
+ * Sends Certificate Signing Requests to config server
+ *
+ * @author freva
+ */
+class CertificateAuthorityClient {
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private final URI caServerUri;
+    private final CloseableHttpClient client;
+
+    CertificateAuthorityClient(String caServerHostname) {
+        this.caServerUri = URI.create("https://" + caServerHostname + ":8443/athenz/v1/provider/sign");
+
+        // Trust all certificates
+        SSLConnectionSocketFactory sslSocketFactory;
+        try {
+            sslSocketFactory = new SSLConnectionSocketFactory(
+                    new SSLContextBuilder().loadTrustMaterial(null, (arg0, arg1) -> true).build(),
+                    SSLConnectionSocketFactory.ALLOW_ALL_HOSTNAME_VERIFIER);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        this.client = HttpClientBuilder.create()
+                .setUserAgent("Node-Maintainer CA client")
+                .setSSLSocketFactory(sslSocketFactory)
+                .disableAutomaticRetries()
+                .build();
+    }
+
+    X509Certificate signCsr(PKCS10CertificationRequest csr) {
+        CsrSerializedPayload csrSerialized = new CsrSerializedPayload(csr);
+        CertificateSerializedPayload certificateSerialized = sendCsr(csrSerialized);
+        return certificateSerialized.certificate;
+    }
+
+    private CertificateSerializedPayload sendCsr(CsrSerializedPayload csrSerialized) {
```

Comment:
Don't (yet) understand why this method doesn't send. Seems like an unnecessary layer.

Reply:
Not ideal naming, but the idea was that `sendCsr()` does serialization/deserialization, while `postJson` does the HTTP stuff, then the try/catch blocks can throw better error message as both of those can throw `IOException`.